src/test/java/com/hatchways/assessment/api/ApplicationTests.java

```diff
@@ -0,0 +1,162 @@
+	}
+
+	// helper function to get the response JSON of the API endpoint
+	public JsonObject getResponseJson(String url) throws IOException {
+		URL siteURL = new URL(url);
+		HttpURLConnection connection = (HttpURLConnection) siteURL.openConnection();
+		int responseCode = connection.getResponseCode();
+		InputStream inputStream;
+		if (200 <= responseCode && responseCode <= 299) {
+			inputStream = connection.getInputStream();
+		} else {
+			inputStream = connection.getErrorStream();
+		}
+		JsonObject json;
+		JsonElement element = new JsonParser().parse(new InputStreamReader(inputStream));
+		json = element.getAsJsonObject();
+		return json;
+	}
+
+	// check if JSON is sorted by its sortingKey
+	public boolean isSorted(JsonObject json, String sortingKey, String direction) {
+		double prev = json.get("posts").getAsJsonArray().get(0).getAsJsonObject().get(sortingKey).getAsDouble();
+		for (JsonElement elem : json.get("posts").getAsJsonArray()) {
+			if (direction.equals("asc")) {
+				if (prev <= elem.getAsJsonObject().get(sortingKey).getAsDouble()) {
+					prev = elem.getAsJsonObject().get(sortingKey).getAsDouble();
+					continue;
+				} else {
+					return false;
+				}
+			}
+			if (direction.equals("desc")) {
+				if (prev >= elem.getAsJsonObject().get(sortingKey).getAsDouble()) {
+					prev = elem.getAsJsonObject().get(sortingKey).getAsDouble();
+				} else {
+					return false;
+				}
+			}
+		}
+		return true;
+	}
+
+	// check if all posts are unique
+	public boolean isUnique(JsonObject json) {
+		HashSet<Integer> set = new HashSet<>();
+		for (JsonElement elem : json.get("posts").getAsJsonArray()) {
+			if (set.contains(elem.getAsJsonObject().get("id").getAsInt())) {
+				return false;
+			} else {
+				set.add(elem.getAsJsonObject().get("id").getAsInt());
+			}
+		}
+		return true;
+	}
+
+	// Test for status code for api/posts route
+	@Test
+	public void testPostAPI() throws Exception {
+		String url = createURLWithPort("/api/posts?tags=tech", "8080");
+		try {
+			assertEquals(200,getStatus(url));
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+	}
+
+	// Test for status code for api/ping route
+	@Test
+	public void testPingAPI() throws Exception {
+		String url = createURLWithPort("/api/ping", "8080");
+		try {
+			assertEquals(200,getStatus(url));
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+	}
+
+	// Test for proper error handling
+	@Test
+	void testErrorResponses() throws Exception {
+		String url = createURLWithPort("/api/posts?tags=tech&sortBy=rubbish", "8080");
+		JsonObject json = getResponseJson(url);
+		assertEquals("{\"error\":\"sortBy parameter is invalid\"}", json.toString());
+		url = createURLWithPort("/api/posts?tags=tech&sortBy=id&direction=rubbish", "8080");
+		json = getResponseJson(url);
+		assertEquals("{\"error\":\"direction parameter is invalid\"}", json.toString());
+	}
+
+	// Test for responses being sorted
+	@Test
+	void testSortedResponses() throws Exception {
```

Comment:
Each of these can be its own test. Better to write tests that are focused and assert one case.

Eg. `testSortingAscById`, and `testSortingDescById`